designdocs/user-handles.md

```diff
@@ -0,0 +1,308 @@
+
+The goal of this proposal is to enable users to create unique handles on the Frequency blockchain using a registry, making it more accessible for users to engage with the network.
+
+## Assumptions
+
+* ```MessageSourceAccount``` exists for a user to create a handle.
+* It is optional for a user to create a handle.
+* Handles are unique and has a 1:1 mapping with ```MessageSourceAccount```.
+* Governance will define the range of suffixes allowed for a given handle.
+* Governance will define the time period a handle must be retired before it can be reused.
+
+## Requirements
+
+The high level requirements for user handles are:
+
+* Support user ability to choose any valid handle and select a suffix from a set of options allowed by the chain.
+* Ensure handles and suffixes are unique and non-conflicting.
+* Make the system resistant to namespace exhaustion, homoglyphs and race conditions.
+* Make the system easy to use and integrate with existing UI and wallet systems.
+
+## Proposal
+
+User handle ```registry``` on frequency chain.
+
+### Handle Guidelines
+
+* User handles must be unique and each msa can only be be mapped to one handle.
+* User handles must be between 2 and 32 characters (32b limit) long.
+* User handles cannot contain the following substrings (or homoglyphs): @, #, :, ., ```
+* User handles cannot be (@admin, @everyone, @all) *blocklist of handles we reserve
+* Suffixes are u32 limited to a range defined by governance
+* Suffix will be randomly constrained.
+* Homoglyph versions of handles should still resolve to the same MSA.
+* After a handle is released, it cannot be reused by a time block set by governance.
+
+### General Steps
+
+* Query the chain for the current seed value.
+* Hash the seed with the user desired handle base.
+* Apply the resulting value as the seed to the specified PRNG.
+* Generate the next 10/20 ```u32``` values from the PRNG these are the possible suffixes within the range of suffixes allowed by the chain.
+* Query the chain to check which values are available, check if the handle with the suffix is available.
+* Choose an available value and attempt to claim it by submitting the full handle with the suffix to the chain.
+* Chain validates the handle and suffix and maps the handle to the ```msa_id```.
+* Chain maintains a reverse mapping from handle to msa for offchain use.
+
+### Chain Steps
+
+* Every 100 blocks (e.g., when block_number % 100 == 1), update the current seed by taking the Merkle root of the previous block. Keep the current and one previous seed so that two seeds, s0 and s1, are part of the chain state at any given time.
```

Comment:
Curious why mod(100) == 1 instead of 0. Is that because we have a lot of things that already happen on those blocks?

Reply:
he reason for choosing mod(100) == 1 instead of mod(100) == 0 is to avoid conflicts with any other system events that may be happening on mod(100) == 0. 

Subject to change: this is what i inferred from Wess' proposal

Reply:
There is already a frame for that https://github.com/paritytech/substrate/tree/master/frame/insecure-randomness-collective-flip this is not a secure random just as they mention it.

Reply:
If seed is not random, the values generated using that seed is also not considered random but it might be fine for our usecase

Reply:
For our use case, we can rely on the network effect for the randomness of the seed. As long as we keep track of both seed-prev and seed-current at any given moment, we can ensure that the range of values generated using the seed will be different.

Reply:
Let me add the details on seed computation from Wess's proposal

Reply:
Added a spike for the same, either use msa_id or some variant of it as seed or use chain for the random seed

Reply:
#1250

Reply:
#1250 , we will probably be spiking on seed and PRNG